### PR TITLE
prefer shani over avx512

### PR DESCRIPTION
**Pull request by arnetheduck**
avx512 is faster than shani but only for full 16-block inputs - for smaller inputs, it degrades quickly and becomes up to 5x slower than shani - prefer shani for now.

* simplify initialization such that `hashtree_init` does not have to be called on supported platforms (and assert explicitly otherwise)
* allow overridring the implementation so that users can provide a "fallback" implementation of the API where hardware support is missing or to override the default auto-detection preferences
* expose auto-detection

### CODE REVIEW

File: src/hashtree.h [hunk 1]
```diff
@@ -23,18 +23,32 @@ SOFTWARE.
...
+/** Perfrom hardware detection and return the fastest function available
+ *
+ * @returns Fastest hash tree function, or NULL if no supported hardware is found
+ */
+hashtree_hash_fcn hashtree_detect();
+
+/** Initialize the library to use the given hash tree function or perform
+ * auto-detection based on the CPU if `NULL` is given.
+ *
+ * @returns 0 if auto-detection failed, non-0 otherwise
```
Comment (potuz) [inline src/hashtree.h:40]: I would prefer typically 0 to mean no failures.

File: src/hashtree.h [hunk 2]
```diff
@@ -23,18 +23,32 @@ SOFTWARE.
...
+ */
+int hashtree_init(hashtree_hash_fcn override);
 
-// Undefined behavior if called before init()
+/* Undefined behavior if called without appropriate hardware support*/
```
Comment (potuz) [inline src/hashtree.h:44]:
```suggestion
/* Undefined behavior if called without appropriate hardware support */
```

File: src/hashtree.h [hunk 3]
```diff
@@ -23,18 +23,32 @@ SOFTWARE.
...
 #include <stdint.h>
 
-void hashtree_init();
+typedef void (*hashtree_hash_fcn)(unsigned char*, const unsigned char*, uint64_t);
+
+/** Perfrom hardware detection and return the fastest function available
+ *
+ * @returns Fastest hash tree function, or NULL if no supported hardware is found
+ */
+hashtree_hash_fcn hashtree_detect();
```
Comment (potuz) [inline src/hashtree.h:35]: what's the usage of this function? I feel like if you're willing to use whatever result this function returns then the previous behavior or `hashtree_init` is enough. If on the other hand the user may want to pass fine grained choices to override `hashtree_init`, then it's better to return here an array of options from an enum that are the supported options.

File: src/test.c [hunk 4]
```diff
@@ -613,12 +609,23 @@ void test_hash_armv8_crypto_multiple_blocks() {
...
+void test_override(unsigned char*, const unsigned char*, uint64_t) {
+    override_called += 1;
+}
+
+void test_hashtree_init() {
+    hashtree_init(test_override);
+
+    unsigned char digest[960];
+    hashtree_hash(digest, test_32_block, 30);
+    TEST_CHECK(override_called);
```
Comment (potuz) [inline src/test.c:622]: `override_called` is unitilialized so this test can give false positivies. Better to initialize it to zero before and also add a `TEST_CHECK(!override_called);` before the call to `hashtree_init`

File: src/hashtree.c [hunk 5]
```diff
@@ -31,45 +33,68 @@ SOFTWARE.
...
+    }
+
+    return !!hash_ptr;
+}
+
 void hashtree_hash(unsigned char *output, const unsigned char *input, uint64_t count) {
     (*hash_ptr)(output, input, count);
 }
+
+static void init_and_hash(unsigned char *output, const unsigned char *input, uint64_t count) {
```
Comment (potuz) [inline src/hashtree.c:95]: this function is not tested.

**Review by potuz — CHANGES_REQUESTED** [review]
LGTM, we can discuss the issue of the enum as a design decision. The only real issue seems to be the unitialized int in the test

File: src/test.c (hunk 4 above)
Comment (arnetheduck) [reply src/test.c:622]: should be fine: https://stackoverflow.com/questions/16015656/are-global-variables-always-initialized-to-zero-in-c

File: src/hashtree.h (hunk 3 above)
Comment (arnetheduck) [reply src/hashtree.h]: The initial idea would be that if I have my own function pointer that I would want to fill with an implementation, I'd use this this call - however, a similar effect can be had from doing `if(!hashtree_init(NULL)) hashtree_init(fallback)`, so its utility is diminished - hidden for now, always easier to expose later than to remove: a4d60e8d7227b6e1b85446a7d471067960e464fb

File: src/test.c (hunk 4 above)
Comment (arnetheduck) [reply src/test.c:624]: check added: a4d60e8d7227b6e1b85446a7d471067960e464fb

File: src/hashtree.h (hunk 1 above)
Comment (arnetheduck) [reply src/hashtree.h:35]: this is the `bool` style which fits neatly in an `if(hashtree_init..)` - if we go with 0, I think we need a `HASHTREE_OK` constant to compare against, wdyt?

File: src/test.c (hunk 4 above)
Comment (potuz) [reply src/test.c:624]: Oh right! I missed this.

**Review by potuz — APPROVED** [review]

**Pull request merged**